### DISCUSSION TITLE
Initialize Y.XMLFragment from HTML/JSON

### DISCUSSION THREAD

**Discussion by SamDuvall** [ueberdosis/tiptap#1427] [General Questions] [answered]
I'm in the process of attempting to convert my Slate editor over to TipTap, because I feel like TipTap is really starting to differentiate itself.

I'm trying to create collaborative TipTap documents where I have the initial HTML/JSON on the server side. I see how to get from HTML to JSON via `@tiptap/html`, but I'm not quite sure how to get from HTML/JSON to a Y.XMLFragment on the server.

Does anyone have a suggestion for how to get a `@tiptap/core` `Editor` up and running on a node.js server? I've been trying to force in JSDOM to appease ProseMirror, but I keep running into issues. I'm also open to any other techniques. Thanks!

**Comment by SamDuvall** [comment, marked as answer]
Ooops. I was accidentally using `generateJSON` from `@tiptap/core` instead of `@tiptap/html`, which was why I was fumbling around with `JSDOM`.

**Reply by hanspagel** [reply]
Have a look at yjs/y-prosemirror, it has helper methods to create XMLFragments. ✌️

**Reply by jloganolson** [reply]
Hey there, I found prosemirrorJSONToYDoc (https://github.com/yjs/y-prosemirror/blob/1c393fb3254cc1ed4933e8326b57c1316793122a/src/lib.js#L221) and presumably we can feed it our JSON/schema for parameters? Is there a way to get just the xmlfragment back vs whole ydoc?
(context: want to create the xmlFragment "offline" when there is no tiptap editor instance)

**Reply by hanspagel** [reply]
Sure. `ydoc.getXmlFragment('content')`

https://github.com/yjs/yjs/blob/master/README.md

BTW, you should have a look at hocuspocus: https://www.hocuspocus.dev/

**Reply by jloganolson** [reply]
I'm a donor to hocuspocus! (albeit a small one) - moved the rest of my message into a fresh thread below so future travelers can see code (and maybe figure it out!)

**Comment by jloganolson** [comment]
Code that generates xmlFragment but tied to a different ydoc - the clone seems busted so unclear how I can do the equivalent of get/set xmlfragment without an instance of the editor (currently use the editor content as a workaround way to get/set but not ideal, especially for offline creation of tiptap xml fragments):
```
import Text from '@tiptap/extension-text'
import OneLineDocument from "@/tiptap-nodes/extension-one-line-document";
import OneLineParagraph from "@/tiptap-nodes/extension-one-line-paragraph";
import { prosemirrorJSONToYDoc } from 'y-prosemirror'

function oneLineToJSON(textContent) {
  return {
    type: "doc",
    content: [
      {
        type: "oneLineParagraph",
        content: [
          {
            type: "text",
            text: textContent,
          },
        ],
      },
    ],
  }
}

export default function oneLineToFragment(frag, id, textContent) {

  const schema = getSchema([
    OneLineDocument,
    OneLineParagraph,
    Text
  ])
  const data = oneLineToJSON(textContent);
  const doc = prosemirrorJSONToYDoc(schema, data, id);
  const fragOrig = doc.getXmlFragment(id);

  frag = fragOrig.clone();
  console.log(fragOrig.toJSON()); //desired fragment
  console.log(frag.toJSON()); //empty?
  return frag;
}
